bigframes/core/compile/polars/compiler.py

```diff
@@ -0,0 +1,333 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import dataclasses
+import functools
+import itertools
+from typing import cast, Sequence, TYPE_CHECKING
+
+import bigframes.core
+import bigframes.core.expression as ex
+import bigframes.core.guid as guid
+import bigframes.core.nodes as nodes
+import bigframes.operations as ops
+import bigframes.operations.aggregations as agg_ops
+
+if TYPE_CHECKING:
+    import polars as pl
+else:
+    try:
+        import polars as pl
+    except Exception:
+        pass
+
+
+@dataclasses.dataclass(frozen=True)
+class PolarsExpressionCompiler:
+    """
+    Simple compiler for converting bigframes expressions to polars expressions.
+
+    Should be extended to dispatch based on bigframes schema types.
+    """
+
+    @functools.singledispatchmethod
```

Comment:
Sounds like this is causing a hard-ish dependency on polars if this module is ever imported. Might be able to exclude the method definition with an if block.